### DISCUSSION TITLE
what individuals to include in tree building

### DISCUSSION THREAD

**Discussion by percyfal** [tskit-dev/tsinfer#488] [Q&A] [answered]
Hi,

I don't know if this is the right forum to post as this is not so much an issue as a question on best practice usage. Is there a mailing list or similar for usage questions?

Anyway, I'm trying out tsinfer on a data set that includes some 20 populations (species / subspecies) and three outgroups. To begin with we are interested in analyzing a population subset (~12 populations). My question is if one should build trees using only the subset or just as well use all of the populations to then subset the trees for the analyses. Also, we have derived the ancestral states based on the outgroups. Should they be excluded from tree building? Is there a best practice way forward?

Cheers,

Per

**Comment by hyanwong** [comment, marked as answer]
Hi Per - good to hear you are trying out tsinfer. @benjeffery has kindly turned on "GitHub discussions" and moved this question there, so hopefully others will also use this as a forum for `tsinfer` questions.

In answer to your question, it's early days for working out "best practice", but normally we would advise putting in as much data as is available. I think this would include the outgrips too, as long as you are reasonably sure that the ancestral state for the entire clade (including the outgroups) is correct.

It would be sensible to exclude groups that are poorly sequenced, but I suspect that it will be individual sites that will be poorly known (or perhaps where the ancestral state is uncertain). These can be excluded from the entire inference process by setting `inference=False` during `add_site()`, or excluded on a per-sample basis by setting some of the data to missing.

**Reply by percyfal** [reply]
Hi Yan, and thanks for the prompt response. It's early days for our testing too but sofar we have only had minor issues that were easy enough to fix. Rather naively I at first included all sites but actually ran out of memory for the largest chromosomes where the number of sites > 1e6 (roughly; the species is highly heterozygous) - on a 4TB RAM machine at that! After rereading your paper I saw that you had indeed built trees from fewer sites (but larger sample sizes). I had implemented similar filtering to that in your paper so in the end I thinned the variants to a reasonable number, but I feel this could be done in a more systematic manner.

It seems the best way forward then is to include everything (230 individuals) for tree building, and then focus on sub-populations of interest in subsequent analyses.

**Reply by hyanwong** [reply]
Hi Per. We've run human chr2 on ~10,000 samples with 4,868,210 sites total (2,639,906 of which are inference sites). That took ~0.43 CPU years and roughly 250 GB memory. So I'm surprised that you ran out of memory. @awohns ran most of this, so he might have some tips.

**Reply by awohns** [reply]
Hi @percyfal, thanks for your question. Yes, I agree with Yan that is surprising that you ran out of memory since we've been running larger datasets using a fraction of 4TB. It would be helpful if you could specify when/where precisely the algorithm died (i.e. in `generate-ancestors`, `match-ancestors`, or in `match-samples`)
edit: also, do you know how many (if any) sites have missing data?

**Reply by percyfal** [reply]
Hi @awohns, that sounds promising! I recall @hyanwong mentioning in his phyloseminar talk the possibility to do large-scale analyses, so yeah, I was also a bit surprised, but it could as well be something in our data (for instance, on that same high-memory machine, we fail to phase two chromosomes with Beagle due to memory(!)) or something with my data-loading scripts. In any case, we shouldn't have any missing data. I believe the algorithm crashes at `generate-ancestors`, but since I didn't save the logs at the time, I'm reproducing it as we speak. Hopefully I'll be able to get back to you tomorrow with more details.

**Reply by percyfal** [reply]
Ok that was quicker than I thought:

```
Sample file created for 202 samples (101 individuals) with 2197683 variable sites.
ga-add   (1/6)100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.20M/2.20M [01:29, 24.5kit/s]
Traceback (most recent call last):
  File "/mnt/nbis-scratch/perun/.snakemake/scripts/tmpvbk5orzl.tsinfer.py", line 147, in <module>
    ts = tsinfer.infer(samples, progress_monitor = True, num_threads=snakemake.threads)
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/inference.py", line 274, in infer
    ancestor_data = generate_ancestors(
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/inference.py", line 366, in generate_ancestors
    generator.run()
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/inference.py", line 939, in run
    self.descriptors = self.ancestor_builder.ancestor_descriptors()
MemoryError

```

I realize now it didn't actually run out of memory; rather, I'm thinking there may be something wrong with the samples data? This code works for most chromosomes though so there shouldn't be a problem with the function calls.

**Reply by erikenbody** [reply]
Hi @awohns  and @hyanwong - I work with Per on this project and am excited about using this method and we are very grateful for your feedback! Just chiming in about the outgroups - in this case the outgroups were used to set the ancestral allele in the first place. We wondered if this might lead them to behave oddly as e.g. they will appear as having mostly the ancestral allele. The thought occurred to us after noticing that the GNN results painted them as having highly heterogeneous ancestry, when I would have expected they have very consistent shared ancestry amongst themselves. However, we realized we were not certain what our null expectation of these groups should be.

**Reply by awohns** [reply]
Hi Erik, glad to hear you're excited about the method and only too happy to help!
Per, I think there might indeed be something wrong with the `SampleData` file if it fails so quickly: `generate-ancestors` is the most memory-efficient step in the process so I'd be really surprised if took anywhere near that much memory. Can you check if the `SampleData` file is finalised? It's as simple as loading up the `SampleData` file in python and seeing if the result of  `yoursampledatafile.finalised` is True. If that's not it, there are a couple other things we could check.
Regarding the outgroups/ancestral alleles @erikenbody, is the pattern you see that the outgroups are heterogeneous amongst themselves, but distinct from the non-outgroups? I'm not sure how ancestral state error could play into this, I think that @hyanwong might be better placed to discuss since he's looked into that before

**Reply by percyfal** [reply]
@awohns I checked the sample data file with `tsinfer ls sampledatafile` and finalised is True. AFAICT, all other values look correct. There could also be an issue with how I have subset the input vcf though, using cyvcf2.VCF.set_samples. I match the sample names to my metadata table, and it works for most chromosomes, but I did notice that for some chromosomes the GNN plots look as if some individuals have been mislabeled. I will add some more checks to make sure this is not the case (the tsinfer script is more or less a copy paste from your tutorial).

**Reply by erikenbody** [reply]
Hi again - thanks @awohns, good question. Our outgroup looks consistently heterogenous, which is what led us to believe it is an artifact. Perhaps of interest is to visualize what we see, so I attach a simple plot of GNN. We study a recent adaptive adaptive radiation consisting of multiple species and many sub populations and as you can see, the tree statistics provide a compelling visualization of their relationships, but the outgroup especially does not perform as I would have predicted. 

![example_gnn](https://user-images.githubusercontent.com/12717495/113895617-ff029e80-977d-11eb-8ddd-ff6b50a07fe6.png)

**Reply by hyanwong** [reply]
Perhaps @jeromekelleher has an idea about the `self.ancestor_builder.ancestor_descriptors()` MemoryError above? That seems like it should be easy to debug, I would have thought. I can have a look over the weekend.

**Reply by jeromekelleher** [reply]
No, I'm not sure what's happened there @hyanwong - I wouldn't have expected that step to consume *that* much memory (This is the [function](https://github.com/tskit-dev/tsinfer/blob/main/_tsinfermodule.c#L236) it's calling into).

@percyfal, would you mind rerunning this with the debug flags turned on? (Something like `-vvv` should turn it all the way up)

**Reply by percyfal** [reply]
@jeromekelleher I make the tsinfer infer call from within the script (based on https://tsinfer.readthedocs.io/en/latest/tutorial.html#reading-a-vcf) but presumably this is what you meant (CLI)?
`tsinfer infer -vvv data.samples --num-threads 150 --progress`
```
2021-04-07 19:47:10,599 [1659] INFO     tsinfer.formats: Loaded SampleData(num_samples=202, num_sites=2197683)
2021-04-07 19:47:10,600 [1659] INFO     tsinfer.formats: Loaded SampleData(num_samples=202, num_sites=2197683)
2021-04-07 19:47:13,895 [1659] DEBUG    tsinfer.inference: Using C AncestorBuilder implementation
2021-04-07 19:47:13,895 [1659] INFO     tsinfer.inference: Starting addition of 2197683 sites
ga-add   (1/6)100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.20M/2.20M [01:39, 22.1kit/s]
2021-04-07 19:48:57,625 [1659] INFO     tsinfer.inference: Finished adding sites
2021-04-07 19:48:57,910 [1659] DEBUG    tsinfer.formats: Syncronously flushing buffer
2021-04-07 19:48:57,911 [1659] DEBUG    tsinfer.formats: Flushing buffer 0: start=0 n=0
2021-04-07 19:48:57,911 [1659] DEBUG    tsinfer.formats: Buffer 0 flush done
2021-04-07 19:48:57,913 [1659] CRITICAL root: Traceback (most recent call last):
  File "/mnt/nbis-scratch/perun/valkyrian/bin/tsinfer", line 8, in <module>
    sys.exit(main())
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/__main__.py", line 5, in main
    cli.tsinfer_main()
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/cli.py", line 499, in tsinfer_main
    args.runner(args)
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/cli.py", line 143, in run_infer
    ts = tsinfer.infer(
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/inference.py", line 274, in infer
    ancestor_data = generate_ancestors(
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/inference.py", line 366, in generate_ancestors
    generator.run()
  File "/mnt/nbis-scratch/perun/valkyrian/lib/python3.9/site-packages/tsinfer/inference.py", line 939, in run
    self.descriptors = self.ancestor_builder.ancestor_descriptors()
MemoryError

```
EDIT: I monitored the actual memory usage and it barely registered 0.1% usage so it doesn't seem to be a matter of exhausting memory resources

**Reply by jeromekelleher** [reply]
This is very strange, thanks for your help in debugging it @percyfal. Do you think you could make the ``data.samples`` file available to us somehow? I think there must a bug of some sort here.

**Reply by percyfal** [reply]
@jeromekelleher Ok I've figured out where the code breaks, but I cannot identify what's causing it. Ideally I would like to extract the region but I would need a pointer from you as how to find it.

In the function [ancestor_builder_finalise](https://github.com/tskit-dev/tsinfer/blob/main/lib/ancestor_builder.c#L506) the variable [focal_sites](https://github.com/tskit-dev/tsinfer/blob/main/lib/ancestor_builder.c#L534) is set to NULL at some point in the iteration through the avl. I printed some information when it failed: 

```
number of samples: 202
focal_sites is NULL
pattern_map->num_sites: 374662
pattern_map->num_samples: 202
tsk_id_t size: 4
pattern_map->num_sites * sizeof(tsk_id_t)): 1498648

1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 0|0 0|0 0|0 0|0 0|0 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1
 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 1|1 2|0 
```

I presume num_sites is the size of the block that is used to build a partial genealogy? Printing it for all regions typically returned single-digit values, although occasionally would jump to a couple of thousand, but never anything this big. At first I thought maybe it was caused by a dangling pointer, but having traversed the sites I don't think so. 

What strikes me as odd though is that the last genotype is coded 2|0 - am I wrong to assume that we only should see 0 (ancestral) / 1 (derived)? I double-checked my samples generation script, but I had correctly made sure only to include biallelic snps coded 0 or 1, so this conversion happens later. Also, this only seems to appear for the failing chromosomes (I checked for the existence of genoypes equalling 2 in a few cases).

I've tried as best I can to narrow down the region where this happens to provide a smaller reproducible example for an issue but I don't directly see how to get the actual chromosome positions from the `pattern_map` struct.

**Reply by hyanwong** [reply]
A [fix](https://github.com/tskit-dev/tsinfer/pull/612) to the memory issue has just been pushed to `main`.
